File: C++/Problems/Problems/problem_3983.cpp
#include "problem_3983.h"

bool problem_3983::test() {
    std::string s = "cat";
    std::string t = "chat";

    bool expected = true;

    auto result = canMakeSubsequence(s, t);

    return result == expected;
}

bool problem_3983::canMakeSubsequence(std::string s, std::string t) {
    int n = t.size();
    int m = s.size();
    if (m > n)
        return false;
    std::vector<int> suf(n);
    int c = 0;
    int j = m - 1;
    for (int i = n - 1; i >= 0; --i)
    {
        suf[i] = c;
        if (j > 0 && t[i] == s[j])
        {
            c++;
            j--;
        }
    }
    j = 0;
    bool changed = false;
    for (int i = 0; i < n && j < m; ++i)
    {
        if (s[j] == t[i])
            j++;
        else if (!changed && suf[i] >= m - j - 1)
        {
            j++;
            changed = true;
        }
    }

    return j == m;
}

File: Python/Problems/Problems/problem3983.py
import heapq
import math
import queue
import sys

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter, deque

from .problem import Problem

class Problem3983(Problem):
    def test(self) -> bool:
        s = "cat"
        t = "chat"

        expected = True

        result = self.canMakeSubsequence(s, t)

        return result == expected

    def canMakeSubsequence(self, s: str, t: str) -> bool:
        n = len(t)
        m = len(s)
        suf = [0] * n
        c = 0
        j = m - 1
        for i in range(n - 1, -1, -1):
            suf[i] = c
            if j > 0 and t[i] == s[j]:
                c += 1
                j -= 1
        j = 0
        i = 0
        changed = False
        while i < n and j < m:
            if s[j] == t[i]:
                j += 1
            elif not changed and suf[i] >= m - j - 1:
                j += 1
                changed = True
            i += 1

        return j == m